### PR TITLE
Feature/fallback gender

### PR DESCRIPTION
**Pull request by melferink**
* fallback for gender determination using Clarity LIMS

### CODE REVIEW

File: ExomeDepth/run_ExomeDepth.py [hunk 1]
```diff
@@ -23,30 +23,46 @@ def valid_read(read):
...
-        elif re.search('[C|P]F', os.path.basename(bam)):
-            print((
-                "Sample {0} has a unknown gender based on chrX reads, but resolved as female based on sampleID"
-            ).format(bam_base))
-            return "female"
-        else:
-            sys.exit("Sample {0} has a unknown gender and will not be analysed".format(bam_base))
+        # Query Clarity for gender
+        gender = get_gender_clarity(bam)
+        if gender == "unknown" and force is True:
```
Comment (ellendejong) [inline ExomeDepth/run_ExomeDepth.py:60]: I would recommend:
`if gender == "unknown" and force:`



https://switowski.com/blog/checking-for-true-or-false/
or 
https://peps.python.org/pep-0008/#programming-recommendations

> Don’t compare boolean values to True or False using ==:
> # Correct:
> if greeting:
> # Wrong:
> if greeting == True:
> Worse:
> 
> # Wrong:
> if greeting is True:

File: ExomeDepth/run_ExomeDepth.py [hunk 2]
```diff
@@ -23,30 +23,46 @@ def valid_read(read):
...
+            gender = settings.force_gender
+            qc_status_gender = "WARNING:GenderForcedTo{}".format(settings.force_gender)
```
Comment (ellendejong) [inline ExomeDepth/run_ExomeDepth.py:62]: Just for your information, not required to change. 
Alternative is using f-strings:
`f"WARNING:GenderForcedTo{settings.force_gender}"`

File: ExomeDepth/run_ExomeDepth.py [hunk 3]
```diff
@@ -23,30 +23,46 @@ def valid_read(read):
...
+        if force is True:
```
Comment (ellendejong) [inline ExomeDepth/run_ExomeDepth.py:63]: see previous comment about check on boolean.

File: ExomeDepth/run_ExomeDepth.py [hunk 4]
```diff
@@ -225,10 +241,11 @@ def call_cnv(args):
     os.makedirs(output_folder, exist_ok=True)
 
     """Determine gender"""
+    qc_status_gender = ''
```
Comment (ellendejong) [inline ExomeDepth/run_ExomeDepth.py:244]: When are you using a single quote or double quote? If unsure, try to adapt previous choice. 
Same is true for other code changes, also in settings file.

**Review by ellendejong — CHANGES_REQUESTED** [review]
Great to see the first unit test in a pull-request, amazing. Nice feature, the parametrize. 

I added a few remarks.

File: ExomeDepth/tests/test_run_ExomeDepth.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python
+import sys
+import os
+
+import pytest
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '../../ExomeDepth'))
```
[the remark this answers is not in the document]
Comment (melferink) [reply ExomeDepth/tests/test_run_ExomeDepth.py:7]: This is the only way I could find to import function get_sampleid (the function to be tested) without using a hardcoded path.

If there are other ways, please let me know :)

File: ExomeDepth/run_ExomeDepth.py [hunk 6]
```diff
@@ -23,30 +23,46 @@ def valid_read(read):
...
+    """Get sample gender from Clarity LIMS."""
+    sample_id = get_sampleid(database.functions.get_sample_id(bam))
+    gender_translation = settings.gender_translation
+    lims_client = Lims(settings.clarity_baseuri, settings.clarity_username, settings.clarity_password)
+    samples = lims_client.get_samples(udf={settings.monster_udf: sample_id})
+    gender_list = []
+    for sample in samples:
+        gender_list.append(gender_translation[sample.udf[settings.geslacht_udf].lower()])
+    if len(set(gender_list)) == 1:
+        return list(set(gender_list))[0]
```
[the remark this answers is not in the document]
Comment (melferink) [reply ExomeDepth/run_ExomeDepth.py:41]: That seems logical, will include (although variable is also not needed anymore)

File: ExomeDepth/settings.py [hunk 7]
```diff
@@ -71,6 +76,12 @@
     '16': 15, '17': 16, '18': 17, '19': 18, '20': 19, '21': 20,
     '22': 21, 'X': 22, 'Y': 23, 'MT': 24
 }
+force_gender = "male"
+
+# Clarity settings
+monster_udp = 'Dx Monsternummer'
+geslacht_udp = 'Dx Geslacht'
```
[the remark this answers is not in the document]
Comment (melferink) [reply ExomeDepth/settings.py:83]: yes! (not sure why this didn't result in a crash/problem while testing)

File: ExomeDepth/run_ExomeDepth.py [hunk 8]
```diff
@@ -23,30 +23,46 @@ def valid_read(read):
...
+
+def get_gender_clarity(bam):
+    """Get sample gender from Clarity LIMS."""
+    sample_id = get_sampleid(database.functions.get_sample_id(bam))
+    gender_translation = settings.gender_translation
+    lims_client = Lims(settings.clarity_baseuri, settings.clarity_username, settings.clarity_password)
+    samples = lims_client.get_samples(udf={settings.monster_udf: sample_id})
+    gender_list = []
+    for sample in samples:
+        gender_list.append(gender_translation[sample.udf[settings.geslacht_udf].lower()])
```
[the remark this answers is not in the document]
Comment (melferink) [reply ExomeDepth/run_ExomeDepth.py:39]: should be udf (typo in setting as you suggested)

File: ExomeDepth/run_ExomeDepth.py (hunk 4 above)
Comment (melferink) [reply ExomeDepth/run_ExomeDepth.py:244]: agree that it is is inconsistent. Changed fully in settings and run_ExomeDepth.py

File: ExomeDepth/run_ExomeDepth.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (melferink) [reply ExomeDepth/run_ExomeDepth.py:39]: After consulting with Robert: Clarity is configured to use man, vrouw, onbekend so its not needed here.

File: ExomeDepth/run_ExomeDepth.py (hunk 1 above)
Comment (melferink) [reply ExomeDepth/run_ExomeDepth.py:60]: Guess it was Wrong and Worse... changed!

**Review by ellendejong — DISMISSED** [review]

**Review by ellendejong — APPROVED** [review]
Typo fix approved.

**Review by rernst — APPROVED** [review]

**Pull request merged**